src/lib/userStorage/UserProfileStorage.js

```diff
@@ -0,0 +1,494 @@
+   * @returns {Promise<{}>}
+   * @private
+   */
+  async _decryptProfileFields(profile): Promise<any> {
+    const outputProfile = {}
+    await Promise.all(
+      Object.keys(profile).map(
+        async item =>
+          typeof profile[item]?.value === 'string' &&
+          (outputProfile[item] = {
+            ...profile[item],
+            value: await this.profiledb.decryptField(profile[item]?.value),
+          }),
+      ),
+    )
+    return outputProfile
+  }
+
+  /**
+   * encrypt decrypted profile
+   * @param {*} profile
+   */
+  async _encryptProfileFields(profile): Promise<any> {
+    let encryptProfile = {}
+    await Promise.all(
+      Object.keys(profile).map(
+        async field =>
+          typeof profile[field]?.value === 'string' &&
+          (encryptProfile[field] = {
+            ...profile[field],
+            value: await this.profiledb.encryptField(profile[field]?.value),
+          }),
+      ),
+    )
+    return encryptProfile
+  }
+
+  /**
+   * saves a complete profile to the underlying storage
+   * @param {*} profile
+   * @param update
+   */
+  async setProfile(profile, update: boolean = false): Promise<any> {
+    if (profile && !profile.validate) {
+      profile = getUserModel(profile)
+    }
+    const fields = Object.keys(profile).filter(prop => UserStorage.indexableFields[prop])
+    let { errors, isValid } = profile.validate(update)
+
+    if (!isValid) {
+      logger.warn(
+        'setProfile failed',
+        'Fields validation failed',
+        new Error('setProfile failed: Fields validation failed'),
+        { errors, category: ExceptionCategory.Human },
+      )
+
+      throw errors
+    }
+
+    const { avatar } = profile
+
+    if (!!avatar && isValidBase64Image(avatar)) {
+      const cids = await this._resizeAndStoreAvatars(avatar)
+
+      assign(profile, cids)
+    }
+
+    const fieldsToSave = fields.reduce(
+      (acc, currKey) => ({
+        ...acc,
+        [currKey]: {
+          value: profile[currKey],
+          display: profile[currKey],
+          privacy: this.getFieldPrivacy(currKey),
+        },
+      }),
+      {},
+    )
+
+    return this.setProfileFields(fieldsToSave)
+  }
+
+  /**
+   * gets the current unencrypted field values in-memory
+   * @returns
+   */
+  getProfile(): Profile {
+    return Object.keys(this.profile).reduce((acc, currKey) => ({ ...acc, [currKey]: this.profile[currKey].value }), {})
+  }
+
+  /**
+   * updates profile fields in storage
+   * @param {*} fields
+   * @returns
+   */
+  async setProfileFields(fields: { [key: string]: ProfileField }): Promise<any> {
+    const encryptedFields = await this._encryptProfileFields(fields)
+
+    await this.profiledb.setProfileFields(encryptedFields)
+    this._setLocalProfile({ ...this.profile, ...fields })
+  }
+
+  /**
+   * Set profile field with privacy
+   * @param {*} field
+   * @param {*} value
+   * @param {*} privacy
+   * @param {*} onlyPrivacy
+   * @returns
+   */
+  setProfileField(
```

Comment:
1. there should be some private internal helper for the display value logic.
2. verify your logic where you safe fields to database specifically setProfile etc that you actually set the display value correctly

Reply:
2. is still not fixed, see new comments